### ISSUE TITLE
Wrong commands' ordering in generated spec files

### ISSUE THREAD

**Issue by homier** [kevinconway/rpmvenv#77]
Hi there!

I found some weird behavior when I tried to build my python package. Without any changes in a configuration file build can randomly fail. After some investigations I decided to look into the generated spec file.
Pay attention to the following output of `rpmvenv --spec` command with the same configuration file:

First run:
```
# Macros
%define venv_cmd virtualenv --always-copy --python=/opt/python38/bin/python3.8
%define venv_name test-venv
%define venv_install_dir /opt/%{venv_name}
%define venv_dir %{buildroot}/%{venv_install_dir}
%define venv_bin %{venv_dir}/bin
%define venv_python %{venv_bin}/python
%define venv_pip %{venv_python} %{venv_bin}/pip install 
%define __prelink_undo_cmd %{nil}
# Globals
%global __os_install_post %(echo '%{__os_install_post}' | sed -e 's!/usr/lib[^[:space:]]*/brp-python-bytecompile[[:space:]].*$!!g')
# Tags
Name: test-rpmvenv-package
Version: 0.1a
Release: 1
BuildRoot: %(mktemp -ud %{_tmppath}/%{SOURCE0}-%{version}-%{release}-XXXXXX)
Requires: python38, systemd
Summary: blahblah
Group: Application/System
License: Proprietary
Url: https://example.com
Source0: /home/homier/Documents/test
AutoReq: No
AutoProv: No
# Blocks
%prep
rm -rf %{buildroot}/*
mkdir -p %{buildroot}/%{venv_install_dir}
%clean
rm -rf %{buildroot}
%install
pathfix.py -pni "%{venv_install_dir}/bin/python3.8 %{py3_shbang_opts}" %{venv_dir}/lib/python3.8/site-packages/ %{venv_bin}/*
%{venv_cmd} %{venv_dir}
cd %{SOURCE0}
%{venv_pip} -r requirements.txt
cd -
find %{venv_dir} -type d -name "__pycache__" -print0 | xargs -0 rm -rf
# RECORD files are used by wheels for checksum. They contain path names which
# match the buildroot and must be removed or the package will fail to build.
find %{buildroot} -name "RECORD" -exec rm -rf {} \;
# Change the virtualenv path to the target installation direcotry.
venvctrl-relocate --source=%{venv_dir} --destination=/%{venv_install_dir}
%description
Blahblah
%files
/%{venv_install_dir}
```

Second run:
```
# Macros
%define venv_cmd virtualenv --always-copy --python=/opt/python38/bin/python3.8
%define venv_name test-venv
%define venv_install_dir /opt/%{venv_name}
%define venv_dir %{buildroot}/%{venv_install_dir}
%define venv_bin %{venv_dir}/bin
%define venv_python %{venv_bin}/python
%define venv_pip %{venv_python} %{venv_bin}/pip install 
%define __prelink_undo_cmd %{nil}
# Globals
%global __os_install_post %(echo '%{__os_install_post}' | sed -e 's!/usr/lib[^[:space:]]*/brp-python-bytecompile[[:space:]].*$!!g')
# Tags
AutoReq: No
AutoProv: No
Name: test-rpmvenv-package
Version: 0.1a
Release: 1
BuildRoot: %(mktemp -ud %{_tmppath}/%{SOURCE0}-%{version}-%{release}-XXXXXX)
Requires: python38, systemd
Summary: blahblah
Group: Application/System
License: Proprietary
Url: https://example.com
Source0: /home/homier/Documents/test
# Blocks
%install
pathfix.py -pni "%{venv_install_dir}/bin/python3.8 %{py3_shbang_opts}" %{venv_dir}/lib/python3.8/site-packages/ %{venv_bin}/*
%{venv_cmd} %{venv_dir}
cd %{SOURCE0}
%{venv_pip} -r requirements.txt
cd -
find %{venv_dir} -type d -name "__pycache__" -print0 | xargs -0 rm -rf
# RECORD files are used by wheels for checksum. They contain path names which
# match the buildroot and must be removed or the package will fail to build.
find %{buildroot} -name "RECORD" -exec rm -rf {} \;
# Change the virtualenv path to the target installation direcotry.
venvctrl-relocate --source=%{venv_dir} --destination=/%{venv_install_dir}
%description
Blahblah
%prep
mkdir -p %{buildroot}/%{venv_install_dir}
rm -rf %{buildroot}/*
%files
/%{venv_install_dir}
%clean
rm -rf %{buildroot}
```
And the third run:
```
# Macros
%define venv_cmd virtualenv --always-copy --python=/opt/python38/bin/python3.8
%define venv_name test-venv
%define venv_install_dir /opt/%{venv_name}
%define venv_dir %{buildroot}/%{venv_install_dir}
%define venv_bin %{venv_dir}/bin
%define venv_python %{venv_bin}/python
%define venv_pip %{venv_python} %{venv_bin}/pip install 
%define __prelink_undo_cmd %{nil}
# Globals
%global __os_install_post %(echo '%{__os_install_post}' | sed -e 's!/usr/lib[^[:space:]]*/brp-python-bytecompile[[:space:]].*$!!g')
# Tags
AutoReq: No
AutoProv: No
Name: test-rpmvenv-package
Version: 0.1a
Release: 1
BuildRoot: %(mktemp -ud %{_tmppath}/%{SOURCE0}-%{version}-%{release}-XXXXXX)
Requires: python38, systemd
Summary: blahblah
Group: Application/System
License: Proprietary
Url: https://example.com
Source0: /home/homier/Documents/test
# Blocks
%prep
mkdir -p %{buildroot}/%{venv_install_dir}
rm -rf %{buildroot}/*
%files
/%{venv_install_dir}
%install
%{venv_cmd} %{venv_dir}
cd %{SOURCE0}
%{venv_pip} -r requirements.txt
cd -
find %{venv_dir} -type d -name "__pycache__" -print0 | xargs -0 rm -rf
# RECORD files are used by wheels for checksum. They contain path names which
# match the buildroot and must be removed or the package will fail to build.
find %{buildroot} -name "RECORD" -exec rm -rf {} \;
# Change the virtualenv path to the target installation direcotry.
venvctrl-relocate --source=%{venv_dir} --destination=/%{venv_install_dir}
pathfix.py -pni "%{venv_install_dir}/bin/python3.8 %{py3_shbang_opts}" %{venv_dir}/lib/python3.8/site-packages/ %{venv_bin}/*
%clean
rm -rf %{buildroot}
%description
Blahblah
```

Look at the `install` and `pre` blocks. The ordering of commands to be executed is completely random. It doesn't make any sense to create a buildroot directory and then delete it immediately. Also since I need to fix shebangs inside python files, I passed pathfix.py command to `{"blocks": "install": []}` expecting it would be executed in the end of installation, not in the beginning.

Environment:
OS - Fedora 31, CentOS 8
Python version - python 3.7, python 3.8

**Comment by kevinconway** [comment]
Your expectations are correct. The order should be deterministic and based on the order of the values in the `extensions.enabled` configuration field. The `%prep` block being out of order may be a hint to the cause. The `rm -rf %{buildroot}/*` line comes from the "core" extension that defines the basic outline of the spec file. All other plugins must come after the core.

Here is the function that loads and orders the extensions:
https://github.com/kevinconway/rpmvenv/blame/38784c4500bfa6233cd8ef1ef8c5fd7c9602fd32/rpmvenv/extensions/loader.py#L22-L31

The `core` value is prepended to the set of enabled plugins. However, I believe the issue you are seeing is due to the `set` usage. The `set` type is not ordered and using it there is most likely causing this bug. I believe this bug has existed since I created this project so it is surprising to hear about it only now. I believe that all combinations of `core`, `python`, and `files` plugins actually work in any order which has masked this issue for several years.

I think the fix may be as simple as modifying this one line: https://github.com/kevinconway/rpmvenv/blame/38784c4500bfa6233cd8ef1ef8c5fd7c9602fd32/rpmvenv/extensions/loader.py#L24. It would be equally valid to either remove the `set` entirely or replace it with `OrderedDict` so that the order is preserved. There also needs to be a test for this so that this issue does happen again. There is a `python_config_file` parameter in the tests that iterates through all combinations of features. A test that generates the spec 20 or 25 times for each input and verifies it is the same as the first rendering should reliably catch this issue.

If you have time to make a PR for this I'll happily review it and help get the fix published. If not then I should have some time this weekend to make a patch.

Thanks for reporting the bug!

**Comment by kevinconway** [comment]
@homier I just released version `0.23.0` which contains a fix for the ordering issues. It will now order based on your `extensions.enabled` list. Thanks for reporting the bug and please re-open the issue if this patch did not fix it.
